### PR TITLE
Update TypeScript/CRA/Jest/etc

### PR DESCRIPTION
**Pull request by jessie-ross**
**This PR upgrades TypeScript, as well as any packages that need to be upgraded because TypeScript was upgraded:**

These unfortunately need to be done in one PR, as the changes are all tied together:
- Upgrades typescript
- Upgrades create-react-app
- Upgrades eslint
- Upgrades jest
- Removes direct references to packages handled by create-react-app

- Fixes new TypeScript errors
- Fixes new eslint issues

**Additionally, this resolves all critical npm security issues.**

Part of :
- https://github.com/Automattic/experimentation-platform/issues/406
- #687

Todo:
- [x] Fix the test infrastructure

## How has this been tested?

- Automated tests that cover added/changed functionality

### CODE REVIEW

**Comment by jessie-ross** [comment]
@razvanpapadopol this is ready for review 🙂

File: src/components/explat/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap [hunk 1]
```diff
@@ -1733,15 +1733,15 @@ exports[`A/B/n: renders correctly for 1 analysis datapoint, not statistically si
                   class="makeStyles-metricValueWrapper-196"
                 >
                   +
-                  355
```
Comment (jessie-ross) [inline src/components/explat/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap:1736]: I'm seeing altered data for 1 analysis datapoint for impact data, not sure what to make of it 🤔

Comment (jessie-ross) [reply src/components/explat/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap:1736]: It's only affecting this one A/B/n non-statistically significant test

Comment (jessie-ross) [reply src/components/explat/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap:1736]: It looks like this is an edge case from having 0 days of results, and isn't a practical issue.

I've compared live results for A/B/n experiments in this branch and in production and the results are identical.

File: src/components/explat/metrics/multi-view/MetricsTable.test.tsx [hunk 2]
```diff
@@ -94,20 +94,20 @@ test('with some metrics, allow searching by metric name and metric params', asyn
...
   await changeFieldByRole('textbox', /Search/, mockMetricName)
   await waitFor(() => expect(screen.queryByText(/metric_1/)).toBeNull())
-  expect(screen.queryByText(mockMetricName)).toBeInTheDocument()
+  screen.getByText(mockMetricName)
 
   // no results
   await changeFieldByRole('textbox', /Search/, 'some other name')
   await waitFor(() => expect(screen.queryByText(mockMetricName)).toBeNull())
-  expect(screen.queryAllByText('1-0 of 0')[0]).toBeInTheDocument()
+  screen.getAllByText('0-0 of 0')
```
Comment (razvanpapadopol) [inline src/components/explat/metrics/multi-view/MetricsTable.test.tsx:107]: Hehe, it's nice to see this little bug fixed 🎉

File: src/components/explat/experiments/wizard/Metrics.tsx [hunk 3]
```diff
@@ -66,6 +66,8 @@ const useStyles = makeStyles((theme: Theme) =>
     },
     metricName: {
       fontFamily: theme.custom.fonts.monospace,
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore; typing bug somewhere
```
Comment (razvanpapadopol) [inline src/components/explat/experiments/wizard/Metrics.tsx:70]: Looks like a bug related to the `makeStyles` helper. When getting the exact same property via `useTheme`, everything is fine.

File: src/components/explat/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap (hunk 1 above)
Comment (razvanpapadopol) [reply src/components/explat/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap:1736]: > edge case from having 0 days of results

I see. Adding some runtime should fix this, making the test more reliable, but we can do that separately.

> I've compared live results for A/B/n experiments in this branch and in production and the results are identical.

Thanks for checking! The demo data also looks fine.

When trying to reduce the date-related unit tests flakiness, I guess this case was left out since it's "hidden" because of the "No results" state of Experiment Results.

**Review by razvanpapadopol — APPROVED** [review]
LGTM! Great work, @jessie-ross, dealing with these critical package updates and their chain of dependencies ❤️
